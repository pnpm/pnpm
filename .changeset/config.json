{
  "$schema": "https://unpkg.com/@changesets/config@1.0.3/schema.json",
  "changelog": "@changesets/cli/changelog",
  "commit": false,
  "linked": [],
  "access": "public",
  "baseBranch": "master"
}
