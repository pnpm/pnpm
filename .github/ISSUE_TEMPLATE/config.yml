blank_issues_enabled: false
contact_links:
  - name: Question 🤷🏻‍♂️
    url: https://github.com/pnpm/pnpm/discussions/new?category=q-a
    about: Ask questions and discuss with other community members
  - name: My company is migrating to pnpm and we have several issues/feature requests
    url: https://github.com/pnpm/pnpm/discussions/3787
