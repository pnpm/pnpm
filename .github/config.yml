# Comment to be posted to on PRs from first time contributors in your repository
newPRWelcomeComment: |
  💖 Thanks for opening this pull request! 💖
  Please be patient and we will get back to you as soon as we can.
# Comment to be posted to on pull requests merged by a first time user
firstPRMergeComment: >
  Congrats on merging your first pull request! 🎉🎉🎉
