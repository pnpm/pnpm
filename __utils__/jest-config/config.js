const path = require('path')

const config = {
  preset: "ts-jest",
  transform: {
    '^.+\\.tsx?$': ['ts-jest', {
      // For most projects, the tsconfig.json and test/tsconfig.json are almost
      // exactly the same. But it's more correct to point to test/tsconfig.json
      // to prevent surprises in the future.
      tsconfig: 'test/tsconfig.json'
    }]
  },
  testMatch: ["**/test/**/*.[jt]s?(x)", "**/src/**/*.test.ts"],
  testEnvironment: "node",
  collectCoverage: true,
  coveragePathIgnorePatterns: ["/node_modules/"],
  testPathIgnorePatterns: ["/fixtures/", "/__fixtures__/", "<rootDir>/test/utils/.+"],
  modulePathIgnorePatterns: ['\/__fixtures__\/.*'],
  testTimeout: 4 * 60 * 1000, // 4 minutes
  setupFilesAfterEnv: [path.join(__dirname, "setupFilesAfterEnv.js")],
  maxWorkers: "50%",
}

if (process.env.PNPM_SCRIPT_SRC_DIR) {
  const pathAsArr = process.env.PNPM_SCRIPT_SRC_DIR.split(path.sep)
  const packageName = pathAsArr[pathAsArr.length - 1]
  config.cacheDirectory = path.join(__dirname, ".jest-cache", packageName)
}

module.exports = config
