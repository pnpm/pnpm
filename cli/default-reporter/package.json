{
  "name": "@pnpm/default-reporter",
  "version": "12.4.8",
  "description": "The default reporter of pnpm",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!*.map"
  ],
  "scripts": {
    "start": "tsc --watch",
    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
    "just-test-preview": "ts-node test --type-check",
    "_test": "jest",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/main/cli/default-reporter",
  "keywords": [
    "pnpm8",
    "pnpm-reporter"
  ],
  "engines": {
    "node": ">=16.14"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "peerDependencies": {
    "@pnpm/logger": "^5.0.0"
  },
  "dependencies": {
    "@pnpm/config": "workspace:*",
    "@pnpm/core-loggers": "workspace:*",
    "@pnpm/dedupe.issues-renderer": "workspace:*",
    "@pnpm/dedupe.types": "workspace:*",
    "@pnpm/error": "workspace:*",
    "@pnpm/render-peer-issues": "workspace:*",
    "@pnpm/types": "workspace:*",
    "ansi-diff": "^1.1.1",
    "boxen": "^5.1.2",
    "chalk": "^4.1.2",
    "cli-truncate": "^2.1.0",
    "normalize-path": "^3.0.0",
    "pretty-bytes": "^5.6.0",
    "pretty-ms": "^7.0.1",
    "ramda": "npm:@pnpm/ramda@0.28.1",
    "right-pad": "^1.0.1",
    "rxjs": "^7.8.1",
    "semver": "^7.5.4",
    "stacktracey": "^2.1.8",
    "string-length": "^4.0.2"
  },
  "devDependencies": {
    "@pnpm/default-reporter": "workspace:*",
    "@types/normalize-path": "^3.0.2",
    "@types/ramda": "0.28.20",
    "@types/semver": "7.5.3",
    "ghooks": "2.0.4",
    "load-json-file": "^6.2.0",
    "normalize-newline": "3.0.0",
    "strip-ansi": "^6.0.1"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/main/cli/default-reporter#readme",
  "funding": "https://opencollective.com/pnpm",
  "exports": {
    ".": "./lib/index.js"
  }
}
