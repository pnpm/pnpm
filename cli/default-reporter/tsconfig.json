{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src",
    "suppressImplicitAnyIndexErrors": true,
    "ignoreDeprecations": "5.0"
  },
  "include": [
    "src/**/*.ts",
    "../../__typings__/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../../config/config"
    },
    {
      "path": "../../packages/core-loggers"
    },
    {
      "path": "../../packages/error"
    },
    {
      "path": "../../packages/render-peer-issues"
    },
    {
      "path": "../../packages/types"
    }
  ],
  "composite": true
}
