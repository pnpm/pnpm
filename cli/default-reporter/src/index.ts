import { type Config } from '@pnpm/config'
import type * as logs from '@pnpm/core-loggers'
import { type LogLevel } from '@pnpm/logger'
import * as Rx from 'rxjs'
import { filter, map, mergeAll } from 'rxjs/operators'
import createDiffer from 'ansi-diff'
import { EOL } from './constants'
import { mergeOutputs } from './mergeOutputs'
import { reporterForClient } from './reporterForClient'
import { formatWarn } from './reporterForClient/utils/formatWarn'
import { reporterForServer } from './reporterForServer'
import { type FilterPkgsDiff } from './reporterForClient/reportSummary'
import { type PeerDependencyRules } from '@pnpm/types'

export { formatWarn }

export function initDefaultReporter (
  opts: {
    useStderr?: boolean
    streamParser: object
    reportingOptions?: {
      appendOnly?: boolean
      logLevel?: LogLevel
      streamLifecycleOutput?: boolean
      aggregateOutput?: boolean
      throttleProgress?: number
      outputMaxWidth?: number
      hideAddedPkgsProgress?: boolean
      hideProgressPrefix?: boolean
      hideLifecycleOutput?: boolean
      hideLifecyclePrefix?: boolean
      peerDependencyRules?: PeerDependencyRules
    }
    context: {
      argv: string[]
      config?: Config
      env?: NodeJS.ProcessEnv
      process?: NodeJS.Process
    }
    filterPkgsDiff?: FilterPkgsDiff
  }
): () => void {
  if (opts.context.argv[0] === 'server') {
    // eslint-disable-next-line
    const log$ = Rx.fromEvent<logs.Log>(opts.streamParser as any, 'data')
    const subscription = reporterForServer(log$, opts.context.config)
    return () => {
      subscription.unsubscribe()
    }
  }
  const outputMaxWidth = opts.reportingOptions?.outputMaxWidth ?? (process.stdout.columns && process.stdout.columns - 2) ?? 80
  const output$ = toOutput$({
    ...opts,
    reportingOptions: {
      ...opts.reportingOptions,
      outputMaxWidth,
    },
  })
  if (opts.reportingOptions?.appendOnly) {
    const writeNext = opts.useStderr
      ? console.error.bind(console)
      : console.log.bind(console)
    const subscription = output$
      .subscribe({
        complete () {}, // eslint-disable-line:no-empty
        error: (err) => {
          console.error(err.message)
        },
        next: writeNext,
      })
    return () => {
      subscription.unsubscribe()
    }
  }
  const diff = createDiffer({
    height: process.stdout.rows,
    outputMaxWidth,
  })
  const subscription = output$
    .subscribe({
      complete () {}, // eslint-disable-line:no-empty
      error: (err) => {
        logUpdate(err.message)
      },
      next: logUpdate,
    })
  const write = opts.useStderr
    ? process.stderr.write.bind(process.stderr)
    : process.stdout.write.bind(process.stdout)
  function logUpdate (view: string) {
    // A new line should always be appended in case a prompt needs to appear.
    // Without a new line the prompt will be joined with the previous output.
    // An example of such prompt may be seen by running: pnpm update --interactive
    if (!view.endsWith(EOL)) view += EOL
    write(diff.update(view))
  }
  return () => {
    subscription.unsubscribe()
  }
}

export function toOutput$ (
  opts: {
    streamParser: object
    reportingOptions?: {
      appendOnly?: boolean
      logLevel?: LogLevel
      outputMaxWidth?: number
      peerDependencyRules?: PeerDependencyRules
      streamLifecycleOutput?: boolean
      aggregateOutput?: boolean
      throttleProgress?: number
      hideAddedPkgsProgress?: boolean
      hideProgressPrefix?: boolean
      hideLifecycleOutput?: boolean
      hideLifecyclePrefix?: boolean
    }
    context: {
      argv: string[]
      config?: Config
      env?: NodeJS.ProcessEnv
      process?: NodeJS.Process
    }
    filterPkgsDiff?: FilterPkgsDiff
  }
): Rx.Observable<string> {
  opts = opts || {}
  const contextPushStream = new Rx.Subject<logs.ContextLog>()
  const fetchingProgressPushStream = new Rx.Subject<logs.FetchingProgressLog>()
  const executionTimePushStream = new Rx.Subject<logs.ExecutionTimeLog>()
  const progressPushStream = new Rx.Subject<logs.ProgressLog>()
  const stagePushStream = new Rx.Subject<logs.StageLog>()
  const deprecationPushStream = new Rx.Subject<logs.DeprecationLog>()
  const summaryPushStream = new Rx.Subject<logs.SummaryLog>()
  const lifecyclePushStream = new Rx.Subject<logs.LifecycleLog>()
  const statsPushStream = new Rx.Subject<logs.StatsLog>()
  const packageImportMethodPushStream = new Rx.Subject<logs.PackageImportMethodLog>()
  const installCheckPushStream = new Rx.Subject<logs.InstallCheckLog>()
  const registryPushStream = new Rx.Subject<logs.RegistryLog>()
  const rootPushStream = new Rx.Subject<logs.RootLog>()
  const packageManifestPushStream = new Rx.Subject<logs.PackageManifestLog>()
  const peerDependencyIssuesPushStream = new Rx.Subject<logs.PeerDependencyIssuesLog>()
  const linkPushStream = new Rx.Subject<logs.LinkLog>()
  const otherPushStream = new Rx.Subject<logs.Log>()
  const hookPushStream = new Rx.Subject<logs.HookLog>()
  const skippedOptionalDependencyPushStream = new Rx.Subject<logs.SkippedOptionalDependencyLog>()
  const scopePushStream = new Rx.Subject<logs.ScopeLog>()
  const requestRetryPushStream = new Rx.Subject<logs.RequestRetryLog>()
  const updateCheckPushStream = new Rx.Subject<logs.UpdateCheckLog>()
  setTimeout(() => {
    opts.streamParser['on']('data', (log: logs.Log) => {
      switch (log.name) {
      case 'pnpm:context':
        contextPushStream.next(log)
        break
      case 'pnpm:execution-time':
        executionTimePushStream.next(log)
        break
      case 'pnpm:fetching-progress':
        fetchingProgressPushStream.next(log)
        break
      case 'pnpm:progress':
        progressPushStream.next(log)
        break
      case 'pnpm:stage':
        stagePushStream.next(log)
        break
      case 'pnpm:deprecation':
        deprecationPushStream.next(log)
        break
      case 'pnpm:summary':
        summaryPushStream.next(log)
        break
      case 'pnpm:lifecycle':
        lifecyclePushStream.next(log)
        break
      case 'pnpm:stats':
        statsPushStream.next(log)
        break
      case 'pnpm:package-import-method':
        packageImportMethodPushStream.next(log)
        break
      case 'pnpm:peer-dependency-issues':
        peerDependencyIssuesPushStream.next(log)
        break
      case 'pnpm:install-check':
        installCheckPushStream.next(log)
        break
      case 'pnpm:registry':
        registryPushStream.next(log)
        break
      case 'pnpm:root':
        rootPushStream.next(log)
        break
      case 'pnpm:package-manifest':
        packageManifestPushStream.next(log)
        break
      case 'pnpm:link':
        linkPushStream.next(log)
        break
      case 'pnpm:hook':
        hookPushStream.next(log)
        break
      case 'pnpm:skipped-optional-dependency':
        skippedOptionalDependencyPushStream.next(log)
        break
      case 'pnpm:scope':
        scopePushStream.next(log)
        break
      case 'pnpm:request-retry':
        requestRetryPushStream.next(log)
        break
      case 'pnpm:update-check':
        updateCheckPushStream.next(log)
        break
      case 'pnpm' as any: // eslint-disable-line
      case 'pnpm:global' as any: // eslint-disable-line
      case 'pnpm:store' as any: // eslint-disable-line
      case 'pnpm:lockfile' as any: // eslint-disable-line
        otherPushStream.next(log)
        break
      }
    })
  }, 0)
  let other = Rx.from(otherPushStream)
  if (opts.context.config?.hooks?.filterLog != null) {
    const filterLogs = opts.context.config.hooks.filterLog
    const filterFn = filterLogs.length === 1
      ? filterLogs[0]
      : (log: logs.Log) => filterLogs.every!((filterLog) => filterLog(log))
    other = other.pipe(filter(filterFn))
  }
  const log$ = {
    context: Rx.from(contextPushStream),
    deprecation: Rx.from(deprecationPushStream),
    fetchingProgress: Rx.from(fetchingProgressPushStream),
    executionTime: Rx.from(executionTimePushStream),
    hook: Rx.from(hookPushStream),
    installCheck: Rx.from(installCheckPushStream),
    lifecycle: Rx.from(lifecyclePushStream),
    link: Rx.from(linkPushStream),
    other,
    packageImportMethod: Rx.from(packageImportMethodPushStream),
    packageManifest: Rx.from(packageManifestPushStream),
    peerDependencyIssues: Rx.from(peerDependencyIssuesPushStream),
    progress: Rx.from(progressPushStream),
    registry: Rx.from(registryPushStream),
    requestRetry: Rx.from(requestRetryPushStream),
    root: Rx.from(rootPushStream),
    scope: Rx.from(scopePushStream),
    skippedOptionalDependency: Rx.from(skippedOptionalDependencyPushStream),
    stage: Rx.from(stagePushStream),
    stats: Rx.from(statsPushStream),
    summary: Rx.from(summaryPushStream),
    updateCheck: Rx.from(updateCheckPushStream),
  }
  const cmd = opts.context.argv[0]
  const outputs: Array<Rx.Observable<Rx.Observable<{ msg: string }>>> = reporterForClient(
    log$,
    {
      appendOnly: opts.reportingOptions?.appendOnly,
      cmd,
      config: opts.context.config,
      env: opts.context.env ?? process.env,
      filterPkgsDiff: opts.filterPkgsDiff,
      peerDependencyRules: opts.reportingOptions?.peerDependencyRules,
      process: opts.context.process ?? process,
      isRecursive: opts.context.config?.['recursive'] === true,
      logLevel: opts.reportingOptions?.logLevel,
      pnpmConfig: opts.context.config,
      streamLifecycleOutput: opts.reportingOptions?.streamLifecycleOutput,
      aggregateOutput: opts.reportingOptions?.aggregateOutput,
      throttleProgress: opts.reportingOptions?.throttleProgress,
      width: opts.reportingOptions?.outputMaxWidth,
      hideAddedPkgsProgress: opts.reportingOptions?.hideAddedPkgsProgress,
      hideProgressPrefix: opts.reportingOptions?.hideProgressPrefix ?? (cmd === 'dlx'),
      hideLifecycleOutput: opts.reportingOptions?.hideLifecycleOutput,
      hideLifecyclePrefix: opts.reportingOptions?.hideLifecyclePrefix,
    }
  )

  if (opts.reportingOptions?.appendOnly) {
    return Rx.merge(...outputs)
      .pipe(
        map((log: Rx.Observable<{ msg: string }>) => log.pipe(map((msg) => msg.msg))),
        mergeAll()
      )
  }
  return mergeOutputs(outputs)
}
