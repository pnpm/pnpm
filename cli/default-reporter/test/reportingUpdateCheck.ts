import { type Config } from '@pnpm/config'
import { updateCheckLogger } from '@pnpm/core-loggers'
import { toOutput$ } from '@pnpm/default-reporter'
import { createStreamParser } from '@pnpm/logger'
import { take } from 'rxjs/operators'
import stripAnsi from 'strip-ansi'

test('does not print update if latest is less than current', (done) => {
  const output$ = toOutput$({
    context: {
      argv: ['install'],
      env: {},
    },
    streamParser: createStreamParser(),
  })

  updateCheckLogger.debug({
    currentVersion: '10.0.0',
    latestVersion: '9.0.0',
  })

  const subscription = output$.subscribe({
    complete: () => done(),
    error: done,
    next: () => {
      done('should not log anything')
    },
  })

  setTimeout(() => {
    done()
    subscription.unsubscribe()
  }, 10)
})

test('print update notification if the latest version is greater than the current', (done) => {
  const output$ = toOutput$({
    context: {
      argv: ['install'],
      config: { recursive: true } as Config,
      env: {},
    },
    streamParser: createStreamParser(),
  })

  updateCheckLogger.debug({
    currentVersion: '10.0.0',
    latestVersion: '11.0.0',
  })

  expect.assertions(1)

  output$.pipe(take(1)).subscribe({
    complete: () => done(),
    error: done,
    next: output => {
      expect(stripAnsi(output)).toMatchSnapshot()
    },
  })
})

test('print update notification for Corepack if the latest version is greater than the current', (done) => {
  const output$ = toOutput$({
    context: {
      argv: ['install'],
      config: { recursive: true } as Config,
      env: {
        COREPACK_ROOT: '/usr/bin/corepack',
      },
    },
    streamParser: createStreamParser(),
  })

  updateCheckLogger.debug({
    currentVersion: '10.0.0',
    latestVersion: '11.0.0',
  })

  expect.assertions(1)

  output$.pipe(take(1)).subscribe({
    complete: () => done(),
    error: done,
    next: output => {
      expect(stripAnsi(output)).toMatchSnapshot()
    },
  })
})

test('print update notification that suggests to use the standalone scripts for the upgrade', (done) => {
  const output$ = toOutput$({
    context: {
      argv: ['install'],
      config: { recursive: true } as Config,
      env: {
        PNPM_HOME: '/home/user/.local/share/pnpm',
      },
      process: {
        pkg: true,
      } as any, // eslint-disable-line
    },
    streamParser: createStreamParser(),
  })

  updateCheckLogger.debug({
    currentVersion: '10.0.0',
    latestVersion: '11.0.0',
  })

  expect.assertions(1)

  output$.pipe(take(1)).subscribe({
    complete: () => done(),
    error: done,
    next: output => {
      expect(stripAnsi(output)).toMatchSnapshot()
    },
  })
})
