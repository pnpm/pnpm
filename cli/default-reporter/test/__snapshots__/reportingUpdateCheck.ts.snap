// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print update notification for Corepack if the latest version is greater than the current 1`] = `
"
   ╭──────────────────────────────────────────────────────────────────╮
   │                                                                  │
   │                Update available! 10.0.0 → 11.0.0.                │
   │   Changelog: https://github.com/pnpm/pnpm/releases/tag/v11.0.0   │
   │         Run "corepack install -g pnpm@11.0.0" to update.         │
   │                                                                  │
   ╰──────────────────────────────────────────────────────────────────╯
"
`;

exports[`print update notification if the latest version is greater than the current 1`] = `
"
   ╭──────────────────────────────────────────────────────────────────╮
   │                                                                  │
   │                Update available! 10.0.0 → 11.0.0.                │
   │   Changelog: https://github.com/pnpm/pnpm/releases/tag/v11.0.0   │
   │                Run "pnpm add -g pnpm" to update.                 │
   │                                                                  │
   ╰──────────────────────────────────────────────────────────────────╯
"
`;

exports[`print update notification that suggests to use the standalone scripts for the upgrade 1`] = `
"
   ╭──────────────────────────────────────────────────────────────────╮
   │                                                                  │
   │                Update available! 10.0.0 → 11.0.0.                │
   │   Changelog: https://github.com/pnpm/pnpm/releases/tag/v11.0.0   │
   │                Run "pnpm self-update" to update.                 │
   │                                                                  │
   ╰──────────────────────────────────────────────────────────────────╯
"
`;
